### PR TITLE
Adding FusedLayerNorm as an algorithm to Composer

### PR DESCRIPTION
**Pull request by moinnadeem**
This PR adds FusedLayerNorm from APEX as an algorithm into Composer.

### CODE REVIEW

File: composer/algorithms/fused_layernorm/fused_layernorm.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+from __future__ import annotations
+
+import logging
+from typing import Optional, Sequence, Union
+
+import torch
+
+try:
+    from apex.normalization.fused_layer_norm import FusedLayerNorm as APEXFusedLayerNorm
```
Comment (hanlint) [inline composer/algorithms/fused_layernorm/fused_layernorm.py:14]: Since we import all algos, this will always trigger, can we do a:

``
try:
   import apex
except ImportError
APEX_INSTALLED = False
```

and then trigger the failures if apply_fused_layernorm is called?

Comment (moinnadeem) [reply composer/algorithms/fused_layernorm/fused_layernorm.py:14]: Ah, good call. Can do!

File: composer/algorithms/fused_layernorm/fused_layernorm.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+    # ensure that the model contains the same d_embed and layernorm.eps throughout
+    for l in [d_embeds, layernorm_eps]:
+        assert len(set(l)) == 1
+
+    d_embed = d_embeds[0]
+    layernorm_eps = layernorm_eps[0]
+
+    # prepare the replacement policy and perform replacement
+    policy = {torch.nn.LayerNorm: lambda x, module_index: APEXFusedLayerNorm(normalized_shape=d_embed, eps=layernorm_eps)}
```
Comment (hanlint) [inline composer/algorithms/fused_layernorm/fused_layernorm.py:53]: As discussed should be possible to create a:
```
def from_LayerNorm(layer: torch.nn.LayerNorm, module_index: int) -> APEXFusedLayerNorm:
    return APEXFUSEDLayerNorm(
        normalized_shape=layer.normalized_shape
        eps=layer.eps
    )
```

and then apply the policy as `{torch.nn.LayerNorm: from_LayerNorm}`

File: tests/fixtures/synthetic_hf_state.py [hunk 3]
```diff
@@ -45,9 +47,8 @@ def make_dummy_lm(model_name: str, max_position_embeddings: int, tokenizer):
 
 
 def make_synthetic_dataloader(dataset_config: dict):
-    """
-    creates a dataloader for synthetic sequence data
-    """
+    """creates a dataloader for synthetic sequence data."""
+    pytest.importorskip('transformers')
```
Comment (jzf2101) [inline tests/fixtures/synthetic_hf_state.py:51]: @hanlint when we put `importtoskip` inside the functions in #992 it skipped the tests in `test_synthetic_hf_state.py` - we may need to come up with a new work around to avoid skipping. currently in this location my tests are skipped.

File: tests/fixtures/synthetic_hf_state.py [hunk 4]
```diff
@@ -31,6 +32,7 @@ def make_lm_tokenizer(config: dict):
 
 
 def make_dummy_lm(model_name: str, max_position_embeddings: int, tokenizer):
+    pytest.importorskip('transformers')
```
Comment (jzf2101) [inline tests/fixtures/synthetic_hf_state.py:35]: skipping here could be a cause as well of `tests/test_synthetic_hf_state.py` being skipped.

File: tests/fixtures/synthetic_hf_state.py [hunk 5]
```diff
@@ -14,15 +14,16 @@
...
     lm_dataset_configs = [
         config[0] for config in test_synthetic_lm_data.generate_parameter_configs(
-            ['num_samples', 'chars_per_sample', 'column_names', 'tokenizer_family'])
+            ['num_samples', 'chars_per_sample', 'column_names', 'tokenizer_family'], model_family=model_family)
     ]
     return lm_dataset_configs
 
 
 def make_lm_tokenizer(config: dict):
+    pytest.importorskip('transformers')
```
Comment (jzf2101) [inline tests/fixtures/synthetic_hf_state.py:26]: skipping here could be a cause as well of `tests/test_synthetic_hf_state.py` being skipped.

**Review by jzf2101 — COMMENTED** [review]
we may need to figure out how to deal with the `pytest.importtoskip(transformers)` statements in `tests/fixtures/synthetic_state.py` because it's currently causing skipping in `tests/test_syntehtic_hf_state.py`

File: tests/fixtures/synthetic_hf_state.py (hunk 3 above)
Comment (hanlint) [reply tests/fixtures/synthetic_hf_state.py:51]: What should the expected behavior be? For jenkins, we always install `.[all]`, and therefore test everything.

File: composer/algorithms/fused_layernorm/fused_layernorm.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+def apply_fused_layernorm(model: torch.nn.Module, optimizers: Union[torch.optim.Optimizer,
+                                                                    Sequence[torch.optim.Optimizer]]) -> None:
+    """Replaces all instances of `torch.nn.LayerNorm` with a `apex.normalization.fused_layer_norm.FusedLayerNorm
+    <https://nvidia.github.io/apex/layernorm.html>`_.
+
+    By fusing multiple kernel launches into one, this usually improves GPU utilization.
+    """
+    if not APEX_INSTALLED:
+        raise ImportError(
+            'https://github.com/NVIDIA/apex is not installed. The Fused LayerNorm algorithm cannot be applied.')
```
Comment (ravi-mosaicml) [inline composer/algorithms/fused_layernorm/fused_layernorm.py:42]: Could these errors be refactored out into a shared helper? I see the message is repeated below.

File: tests/algorithms/algorithm_settings.py [hunk 7]
```diff
@@ -11,17 +11,31 @@
...
+                                 NoOpModel, ProgressiveResizing, RandAugment, SelectiveBackprop, SeqLengthWarmup,
+                                 SqueezeExcite, StochasticDepth)
 from composer.models import ComposerResNet
 from composer.models.base import ComposerModel
 from tests import common
+from tests.fixtures.synthetic_hf_state import (make_synthetic_bert_dataloader, make_synthetic_bert_model,
+                                               make_synthetic_gpt2_dataloader, make_synthetic_gpt2_model)
+
+simple_bert_settings = {
+    'model': make_synthetic_bert_model(),
```
Comment (ravi-mosaicml) [inline tests/algorithms/algorithm_settings.py:29]: Instead of putting the constructed model in the dictionary, and `model` be just the function?
Then, `get_alg_model` will construct it.

Same thing re: dataloader and below for `simpel_gpt2_settings`.

File: tests/algorithms/algorithm_settings.py [hunk 8]
```diff
@@ -145,21 +160,30 @@ def get_alg_kwargs(alg_cls: Type[Algorithm]) -> Dict[str, Any]:
 def get_alg_model(alg_cls: Type[Algorithm]) -> ComposerModel:
     """Return an instance of the model for an algorithm."""
     settings = _get_alg_settings(alg_cls)['model']
+    if isinstance(settings, ComposerModel):
```
Comment (ravi-mosaicml) [inline tests/algorithms/algorithm_settings.py:163]: Can you remove this check? See above comments re: just putting the constructor in the dictionary.

File: composer/algorithms/fused_layernorm/README.md [hunk 9]
```diff
@@ -0,0 +1,70 @@
+# 🍰 Fused LayerNorm
```
Comment (ravi-mosaicml) [inline composer/algorithms/fused_layernorm/README.md:1]: Thoughts about using the fondue emoij? lol
🫕

Comment (moinnadeem) [reply composer/algorithms/fused_layernorm/README.md:1]: Why fondue? Because it is fused? 

I'm not very opinionated on the emoji... I chose cake because of the "layers", and "LayerNorm"

I also defer to the emoji master himself, @hanlint

Comment (ravi-mosaicml) [reply composer/algorithms/fused_layernorm/README.md:1]: Founde is very fused haha; I'm missing the pun with cake.

File: tests/fixtures/synthetic_hf_state.py (hunk 3 above)
Comment (ravi-mosaicml) [reply tests/fixtures/synthetic_hf_state.py:51]: We need the importskips, as we have a Jenkins install that tests on just a dev install (this check was fixed in #1163)

File: tests/algorithms/test_fused_layernorm.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+        if isinstance(module_class, APEXFusedLayerNorm):
+            found_apex_layernorm = True
+
+    assert found_apex_layernorm
+
+
+@device('gpu')
+def test_fused_layernorm_functional(synthetic_bert_state: Tuple, device: str):
+    state, model, _ = synthetic_bert_state
+    print('Model:', model)
```
Comment (ravi-mosaicml) [inline tests/algorithms/test_fused_layernorm.py:41]:
```suggestion
```

File: tests/algorithms/test_fused_layernorm.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+
+def assert_is_fln_instance(model: BERTModel):
+    pytest.importorskip('apex')
+    from apex.normalization.fused_layer_norm import FusedLayerNorm as APEXFusedLayerNorm
+
+    # ensure that within the entire model, no PyTorch LayerNorm exists, and at least one APEX FLN does.
+    found_apex_layernorm = False
+    for module_class in model.modules():
+        if isinstance(module_class, LayerNorm):
```
Comment (ravi-mosaicml) [inline tests/algorithms/test_fused_layernorm.py:30]: nit: `Exception` and `BaseException` should never be raised directly; here in tests an assertion error is fine

```suggestion
        assert not isinstance(module_class, LayerNorm), 'A torch.nn.LayerNorm ...'
```

File: tests/algorithms/test_fused_layernorm.py [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+    synthetic_config = make_dataset_configs(model_family=['bert'])[0]
+    return synthetic_hf_state_maker(synthetic_config)
+
+
+def assert_is_fln_instance(model: BERTModel):
+    pytest.importorskip('apex')
+    from apex.normalization.fused_layer_norm import FusedLayerNorm as APEXFusedLayerNorm
+
+    # ensure that within the entire model, no PyTorch LayerNorm exists, and at least one APEX FLN does.
+    found_apex_layernorm = False
```
Comment (ravi-mosaicml) [inline tests/algorithms/test_fused_layernorm.py:28]:
```suggestion
    assert any(isinstance(module_class, APEXFusedLayerNorm) for x in model.modules()), "Fused Layernorm not found"
```

File: tests/algorithms/algorithm_settings.py [hunk 13]
```diff
@@ -159,14 +174,26 @@ def get_alg_model(alg_cls: Type[Algorithm]) -> ComposerModel:
...
+    else:
+        raise ValueError(f'Neither dataset nor dataloader have been provided for algorithm {alg_cls}')
+
     if isinstance(settings, tuple):
         (cls, kwargs) = settings
     else:
         (cls, kwargs) = (settings, {})
-    return cls(**kwargs)
+
+    data_cls = cls(**kwargs)
```
Comment (ravi-mosaicml) [inline tests/algorithms/algorithm_settings.py:193]:
```suggestion
    dataloader = cls(**kwargs)
```

File: tests/algorithms/algorithm_settings.py [hunk 14]
```diff
@@ -159,14 +174,26 @@ def get_alg_model(alg_cls: Type[Algorithm]) -> ComposerModel:
...
+    if isinstance(data_cls, Dataset):
+        data_cls = DataLoader(dataset=data_cls, batch_size=4)
```
Comment (ravi-mosaicml) [inline tests/algorithms/algorithm_settings.py:195]:
```suggestion
        dataloader = DataLoader(dataset=dataloader, batch_size=4)
```

File: tests/algorithms/algorithm_settings.py [hunk 15]
```diff
@@ -159,14 +174,26 @@ def get_alg_model(alg_cls: Type[Algorithm]) -> ComposerModel:
...
+    return data_cls
```
Comment (ravi-mosaicml) [inline tests/algorithms/algorithm_settings.py:196]:
```suggestion
    return dataloader
```

File: composer/algorithms/fused_layernorm/fused_layernorm.py [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+    <https://nvidia.github.io/apex/layernorm.html>`_.
+
+    By fusing multiple kernel launches into one, this usually improves GPU utilization.
+    """
+    check_if_apex_installed()
+
+    # prepare the replacement policy and perform replacement
+    policy: Dict[Type[torch.nn.Module], module_surgery.ReplacementFunction] = {torch.nn.LayerNorm: from_LayerNorm}
+    replaced_instances = module_surgery.replace_module_classes(module=model, optimizers=optimizers, policies=policy)
+    log.info(f'Successfully replaced {len(replaced_instances)} of LayerNorm with a Fused LayerNorm.')
```
Comment (ravi-mosaicml) [inline composer/algorithms/fused_layernorm/fused_layernorm.py:52]: Could you emit the `NoEffectWarning` if no replacements were made? https://github.com/mosaicml/composer/blob/0f1624ed3d36aa7c032209f115b3af9ec9e0402d/composer/algorithms/warnings.py#L5

**Review by ravi-mosaicml — APPROVED** [review]
LGTM! See comments.

File: composer/algorithms/fused_layernorm/README.md (hunk 9 above)
Comment (hanlint) [reply composer/algorithms/fused_layernorm/README.md:1]: Cake is good :)

File: composer/yamls/algorithms/fused_layernorm.yaml [hunk 17]
```diff
@@ -0,0 +1,2 @@
+# Copyright 2022 MosaicML Composer authors
```
Comment (hanlint) [inline composer/yamls/algorithms/fused_layernorm.yaml:1]: if this is empty, is it needed?

Comment (moinnadeem) [reply composer/yamls/algorithms/fused_layernorm.yaml:1]: The existence of the YAML file is needed for things like `test_alg_yamls`, since it finds the algorithms to test (and their default hyperparameters) from there. This file happens to be empty because FLN doesn't have any hyperparameters.
